File: src-tauri/src/commands/vfx_session.rs
/* VFX porting session commands — open a skin bin plus every bin its `linked:`
list resolves to into one resident multi-bin session, port systems and
emitters between sessions, and edit idle / persistent / resolver data
natively. The heavy lifting lives in `quartz_lib::vfx_session`; these
commands stay thin and translate to/from the camelCase shapes the frontend
consumes. */

use quartz_lib::vfx_session::construct::{ChildParams, PersistentPayload};
use quartz_lib::vfx_session::ops::{self, PortEmittersResult, PortSystemResult};
use quartz_lib::vfx_session::path::VfxPath;
use quartz_lib::vfx_session::project::VfxPortModel;
use quartz_lib::vfx_session::session;
use serde::Serialize;

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct VfxOpenResult {
    pub session_id: u64,
    pub model: VfxPortModel,
}

/// Open a skin bin (and its resolvable linked bins) into a resident session.
/// Runs on the blocking pool — parsing several bins shouldn't stall the async
/// runtime.
#[tauri::command]
pub async fn vfx_open(path: String) -> Result<VfxOpenResult, String> {
    tokio::task::spawn_blocking(move || session::open(&path))
        .await
        .map_err(|e| format!("Open task failed to join: {}", e))?
        .map(|r| VfxOpenResult {
            session_id: r.session_id,
            model: r.model,
        })
        .map_err(|e| e.to_string())
}

/// Reproject the model from the live trees.
#[tauri::command]
pub async fn vfx_model(session_id: u64) -> Result<VfxPortModel, String> {
    tokio::task::spawn_blocking(move || session::model_of(session_id))
        .await
        .map_err(|e| format!("Model task failed to join: {}", e))?
        .map_err(|e| e.to_string())
}

/// Reparse the session if any source BIN changed outside Quartz.
#[tauri::command]
pub async fn vfx_reload_if_changed(session_id: u64) -> Result<Option<VfxPortModel>, String> {
    tokio::task::spawn_blocking(move || session::reload_if_changed(session_id))
        .await
        .map_err(|e| format!("Reload task failed to join: {}", e))?
        .map_err(|e| e.to_string())
}

/// Write every dirty bin back to its own original file. Returns the paths
/// written. Unless `force` is true, a bin whose file changed on disk since
/// opening aborts the save with a `STALE_FILE:` error so the UI can prompt.
#[tauri::command]
pub async fn vfx_save(session_id: u64, force: Option<bool>) -> Result<Vec<String>, String> {
    let force = force.unwrap_or(false);
    tokio::task::spawn_blocking(move || session::save(session_id, force))
        .await
        .map_err(|e| format!("Save task failed to join: {}", e))?
        .map(|paths| {
            paths
                .iter()
                .map(|p| p.to_string_lossy().into_owned())
                .collect()
        })
        .map_err(|e| e.to_string())
}

/// Close a session and free its trees. Idempotent.
#[tauri::command]
pub async fn vfx_close(session_id: u64) -> Result<bool, String> {
    Ok(session::close(session_id))
}

/// Undo the last edit. Returns the refreshed model, or null if nothing to undo.
#[tauri::command]
pub async fn vfx_undo(session_id: u64) -> Result<Option<VfxPortModel>, String> {
    tokio::task::spawn_blocking(move || session::undo(session_id))
        .await
        .map_err(|e| format!("Undo task failed to join: {}", e))?
        .map_err(|e| e.to_string())
}

/// Redo the last undone edit. Returns the refreshed model, or null if nothing
/// to redo.
#[tauri::command]
pub async fn vfx_redo(session_id: u64) -> Result<Option<VfxPortModel>, String> {
    tokio::task::spawn_blocking(move || session::redo(session_id))
        .await
        .map_err(|e| format!("Redo task failed to join: {}", e))?
        .map_err(|e| e.to_string())
}

/// Create a new empty VFX system named `name` and register it in the resolver.
#[tauri::command]
pub async fn vfx_create_system(session_id: u64, name: String) -> Result<VfxPortModel, String> {
    tokio::task::spawn_blocking(move || ops::create_system(session_id, &name))
        .await
        .map_err(|e| format!("Create system task failed to join: {}", e))?
        .map_err(|e| e.to_string())
}

/// Clone emitters from a donor session into a target system. `assetPaths` are
/// the donor asset strings the frontend copies via port_copy_assets_to_target.
#[tauri::command]
pub async fn vfx_port_emitters(
    target_session_id: u64,
    donor_session_id: u64,
    donor_emitters: Vec<VfxPath>,
    target_system: VfxPath,
) -> Result<PortEmittersResult, String> {
    tokio::task::spawn_blocking(move || {
        ops::port_emitters(
            target_session_id,
            donor_session_id,
            &donor_emitters,
            &target_system,
        )
    })
    .await
    .map_err(|e| format!("Port emitters task failed to join: {}", e))?
    .map_err(|e| e.to_string())
}

/// Clone a whole system from a donor session into the target, renaming unless
/// `preserveName` is set, and upsert it into the resolver.
#[tauri::command]
pub async fn vfx_port_system(
    target_session_id: u64,
    donor_session_id: u64,
    donor_system: VfxPath,
    desired_name: Option<String>,
    preserve_name: bool,
) -> Result<PortSystemResult, String> {
    tokio::task::spawn_blocking(move || {
        ops::port_system(
            target_session_id,
            donor_session_id,
            &donor_system,
            desired_name.as_deref(),
            preserve_name,
        )
    })
    .await
    .map_err(|e| format!("Port system task failed to join: {}", e))?
    .map_err(|e| e.to_string())
}

/// Remove one emitter from its owning system.
#[tauri::command]
pub async fn vfx_delete_emitter(session_id: u64, emitter: VfxPath) -> Result<VfxPortModel, String> {
    tokio::task::spawn_blocking(move || ops::delete_emitter(session_id, &emitter))
        .await
        .map_err(|e| format!("Delete emitter task failed to join: {}", e))?
        .map_err(|e| e.to_string())
}

/// Remove several emitters as one edit (one undo step, one reprojection).
#[tauri::command]
pub async fn vfx_delete_emitters(
    session_id: u64,
    emitters: Vec<VfxPath>,
) -> Result<VfxPortModel, String> {
    tokio::task::spawn_blocking(move || ops::delete_emitters(session_id, &emitters))
        .await
        .map_err(|e| format!("Delete emitters task failed to join: {}", e))?
        .map_err(|e| e.to_string())
}

/// Remove a system entry and any resolver entries pointing at it.
#[tauri::command]
pub async fn vfx_delete_system(session_id: u64, system: VfxPath) -> Result<VfxPortModel, String> {
    tokio::task::spawn_blocking(move || ops::delete_system(session_id, &system))
        .await
        .map_err(|e| format!("Delete system task failed to join: {}", e))?
        .map_err(|e| e.to_string())
}

/// Upsert (16 floats) or remove (null) a system's `transform` matrix.
#[tauri::command]
pub async fn vfx_set_matrix(
    session_id: u64,
    system: VfxPath,
    values: Option<Vec<f32>>,
) -> Result<VfxPortModel, String> {
    let values = match values {
        Some(v) => Some(
            <[f32; 16]>::try_from(v)
                .map_err(|_| "Matrix must have exactly 16 values".to_string())?,
        ),
        None => None,
    };
    tokio::task::spawn_blocking(move || ops::set_matrix(session_id, &system, values))
        .await
        .map_err(|e| format!("Set matrix task failed to join: {}", e))?
        .map_err(|e| e.to_string())
}

/// Add one idle-particle effect per bone for `effectKey`.
#[tauri::command]
pub async fn vfx_idle_add(
    session_id: u64,
    effect_key: String,
    bones: Vec<String>,
) -> Result<VfxPortModel, String> {
    tokio::task::spawn_blocking(move || ops::idle_add(session_id, &effect_key, &bones))
        .await
        .map_err(|e| format!("Idle add task failed to join: {}", e))?
        .map_err(|e| e.to_string())
}

/// Remove every idle-particle effect matching `effectKey`.
#[tauri::command]
pub async fn vfx_idle_remove(session_id: u64, effect_key: String) -> Result<VfxPortModel, String> {
    tokio::task::spawn_blocking(move || ops::idle_remove(session_id, &effect_key))
        .await
        .map_err(|e| format!("Idle remove task failed to join: {}", e))?
        .map_err(|e| e.to_string())
}

/// Append a new child (_cbdl) emitter to the host system.
#[tauri::command]
pub async fn vfx_child_add(
    session_id: u64,
    host_system: VfxPath,
    params: ChildParams,
) -> Result<VfxPortModel, String> {
    tokio::task::spawn_blocking(move || ops::child_add(session_id, &host_system, &params))
        .await
        .map_err(|e| format!("Child add task failed to join: {}", e))?
        .map_err(|e| e.to_string())
}

/// Overwrite an existing child emitter's parameters.
#[tauri::command]
pub async fn vfx_child_update(
    session_id: u64,
    emitter: VfxPath,
    params: ChildParams,
) -> Result<VfxPortModel, String> {
    tokio::task::spawn_blocking(move || ops::child_update(session_id, &emitter, &params))
        .await
        .map_err(|e| format!("Child update task failed to join: {}", e))?
        .map_err(|e| e.to_string())
}

/// Append (index = null) or replace (index = i) a persistent-effect condition.
#[tauri::command]
pub async fn vfx_persistent_upsert(
    session_id: u64,
    index: Option<u32>,
    payload: PersistentPayload,
) -> Result<VfxPortModel, String> {
    tokio::task::spawn_blocking(move || {
        ops::persistent_upsert(session_id, index.map(|i| i as usize), &payload)
    })
    .await
    .map_err(|e| format!("Persistent upsert task failed to join: {}", e))?
    .map_err(|e| e.to_string())
}

/// Remove one persistent-effect condition by index.
#[tauri::command]
pub async fn vfx_persistent_remove(session_id: u64, index: u32) -> Result<VfxPortModel, String> {
    tokio::task::spawn_blocking(move || ops::persistent_remove(session_id, index as usize))
        .await
        .map_err(|e| format!("Persistent remove task failed to join: {}", e))?
        .map_err(|e| e.to_string())
}

/// Upsert a resolver map entry (key hash -> system link).
#[tauri::command]
pub async fn vfx_resolver_upsert(
    session_id: u64,
    key: String,
    value: String,
) -> Result<VfxPortModel, String> {
    tokio::task::spawn_blocking(move || ops::resolver_upsert_op(session_id, &key, &value))
        .await
        .map_err(|e| format!("Resolver upsert task failed to join: {}", e))?
        .map_err(|e| e.to_string())
}

/// Set an emitter's `emitterName`.
#[tauri::command]
pub async fn vfx_rename_emitter(
    session_id: u64,
    emitter: VfxPath,
    new_name: String,
) -> Result<VfxPortModel, String> {
    tokio::task::spawn_blocking(move || ops::rename_emitter(session_id, &emitter, &new_name))
        .await
        .map_err(|e| format!("Rename emitter task failed to join: {}", e))?
        .map_err(|e| e.to_string())
}

/// Rewrite an emitter's texture path. `old_path` identifies the texture node
/// (its current value); `new_path` replaces it.
#[tauri::command]
pub async fn vfx_set_texture(
    session_id: u64,
    emitter: VfxPath,
    old_path: String,
    new_path: String,
) -> Result<VfxPortModel, String> {
    tokio::task::spawn_blocking(move || {
        ops::set_texture(session_id, &emitter, &old_path, &new_path)
    })
    .await
    .map_err(|e| format!("Set texture task failed to join: {}", e))?
    .map_err(|e| e.to_string())
}

/// Rename a system (particleName/particlePath + path hash + resolver relink).
#[tauri::command]
pub async fn vfx_rename_system(
    session_id: u64,
    system: VfxPath,
    new_name: String,
) -> Result<VfxPortModel, String> {
    tokio::task::spawn_blocking(move || ops::rename_system(session_id, &system, &new_name))
        .await
        .map_err(|e| format!("Rename system task failed to join: {}", e))?
        .map_err(|e| e.to_string())
}

File: src/lib/api/vfxSession.ts
import { invokeCommand } from './core';
import type { Step } from './bineditor';

/* VFX porting sessions — a skin bin plus every bin its `linked:` list resolves
   to is opened once into a resident multi-bin Rust session. The frontend holds
   only the structured VfxPortModel projection and edits via commands that
   mutate the owning tree in place; save writes back only the dirty bins. */

export type { Step };

/** Opaque address of a node inside a session's bin trees; pass back verbatim. */
export interface VfxPath {
    bin: number;
    entry: number;
    steps: Step[];
}

export interface BinInfo {
    path: string;
    fileName: string;
    role: 'main' | 'linked';
    dirty: boolean;
}

export interface ChildView {
    effectKey: string;
    rate: number;
    lifetime: number;
    bindWeight: number;
    translation: [number, number, number];
    isSingleParticle: boolean;
    timeBeforeFirstEmission: number;
}

export interface PortEmitter {
    /** `${systemKey}__emitter_${idx}` across both emitter lists. */
    key: string;
    /** emitterName or "Unnamed". */
    name: string;
    complex: boolean;
    isChild: boolean;
    textures: string[];
    meshes: string[];
    /** Display swatches: constants plus up to 6 keyframe colors. */
    colors: [number, number, number, number][];
    path: VfxPath;
    childData: ChildView | null;
}

export interface PortSystem {
    /** Hex path_hash, suffixed "@{bin}" on cross-bin collision. */
    key: string;
    name: string;
    particleName: string | null;
    particlePath: string | null;
    binIndex: number;
    path: VfxPath;
    /** 16 floats when the system has a `transform` field. */
    transform: number[] | null;
    emitters: PortEmitter[];
}

export interface ResolverEntryView {
    key: string;
    value: string;
}

export interface ResolverView {
    binIndex: number;
    entryIndex: number;
    entries: ResolverEntryView[];
}

export interface IdleView {
    effectKey: string;
    bones: string[];
}

export interface PersistentPresetView {
    type: string;
    animationName: string | null;
    scriptName: string | null;
    spellHash: string | null;
    slot: number | null;
    operator: number | null;
    value: number | null;
    delayOn: number;
    delayOff: number;
    /** Unknown driver preserved verbatim. */
    raw: boolean;
}

export interface PersistentVfxView {
    key: string;
    boneName: string | null;
    scale: number | null;
    ownerOnly: boolean | null;
    attachToCamera: boolean | null;
    forceRender: boolean | null;
}

export interface PersistentView {
    index: number;
    label: string;
    preset: PersistentPresetView;
    vfx: PersistentVfxView[];
    submeshesShow: string[];
    submeshesHide: string[];
}

export interface EffectKeyOption {
    key: string;
    label: string;
    particleName: string | null;
}

export interface VfxPortModel {
    /** Identity of the session tree these paths index into. Echo it back on a
     *  cross-session port so paths from a reloaded donor are rejected. */
    generation: number;
    bins: BinInfo[];
    systems: PortSystem[];
    resolver: ResolverView | null;
    idle: IdleView[];
    persistent: PersistentView[];
    effectKeys: EffectKeyOption[];
    submeshes: string[];
    /** True when the main bin has SkinCharacterDataProperties (idle/persistent ops need it). */
    hasSkinCharacterData: boolean;
}

/* ── Animation read layer ────────────────────────────────────────────────────
   The clip graph, its blend masks and tracks, and the skeleton those masks
   index into. Read-only: these shapes mirror `quartz_lib::anim_graph` and
   `vfx_session::anm`, which model the fields the viewer needs and NOT every
   field a clip can carry — every node keeps a `BinAddr` so a write path can
   reach the unmodelled ones without reconstructing the node. */

/** Same wire shape as {@link VfxPath}: where a parsed node lives in the bins.
 *  A map hop (clip / event maps) is recorded as the entry's POSITION. */
export interface BinAddr {
    bin: number;
    entry: number;
    steps: Step[];
}

/** While `[startFrame, endFrame]` is live, hide `hide` and show `show`.
 *  Tokens are submesh names, or `0x`-hashes the hash DB could not resolve. */
export interface SubmeshVisEvent {
    startFrame: number | null;
    endFrame: number | null;
    show: string[];
    hide: string[];
}

/** A `mBoneName` / `mTargetBoneName` pair on a particle event. It has no name
 *  and no hash of its own, so `addr` is the only way to address one. */
export interface ParticlePair {
    boneName: string | null;
    targetBoneName: string | null;
    addr: BinAddr;
}

/** Typed payload of one clip event; switch on `type`. `unknown` preserves an
 *  unmodelled event class with its hash rather than dropping it. */
export type AnimEventKind =
    | {
          type: 'particle';
          effectKey: string | null;
          startFrame: number | null;
          isLoop: boolean | null;
          pairs: ParticlePair[];
      }
    | { type: 'sound'; soundName: string | null; isLoop: boolean | null }
    | {
          type: 'submeshVisibility';
          startFrame: number | null;
          endFrame: number | null;
          show: string[];
          hide: string[];
      }
    | { type: 'faceTarget'; endFrame: number | null; yRotationDegrees: number | null }
    | {
          type: 'conformToPath';
          maskDataName: string | null;
          blendInTime: number | null;
          blendOutTime: number | null;
      }
    | {
          type: 'lockRootOrientation';
          startFrame: number | null;
          endFrame: number | null;
          jointName: string | null;
          blendOutTime: number | null;
      }
    | { type: 'stopAnimation'; stopAnimationName: string | null }
    | { type: 'unknown'; classHash: number };

export interface AnimEvent {
    /** Resolved map key, or `0x{hash}` when the hash DB could not name it. */
    name: string;
    classHash: number;
    kind: AnimEventKind;
    addr: BinAddr;
}

/** One entry of a selector / parametric / condition-float pair list. */
export interface ClipPair {
    clipName: string | null;
    probability: number | null;
    value: number | null;
    addr: BinAddr;
}

/** Which class a clip is, with its class-specific payload; switch on `type`. */
export type ClipKind =
    | { type: 'atomic' }
    | { type: 'sequencer' }
    | { type: 'parallel' }
    | { type: 'selector' }
    | { type: 'parametric' }
    | { type: 'conditionFloat' }
    | { type: 'conditionBool'; trueClip: string | null; falseClip: string | null }
    | { type: 'unknown' };

/** One member of a composite clip's queue: a leaf clip naming a `.anm`. */
export interface ClipMember {
    name: string;
    anmPath: string;
    /** Legacy submesh-only view; a strict subset of `allEvents`. */
    events: SubmeshVisEvent[];
    loops: boolean;
    kind: ClipKind;
    classHash: number;
    /** Raw `mFlags`, unmasked. `loops` is bit 2 of this. */
    flags: number;
    trackDataName: string | null;
    maskDataName: string | null;
    startFrame: number | null;
    endFrame: number | null;
    allEvents: AnimEvent[];
    addr: BinAddr;
}

export interface ClipInfo {
    name: string;
    /** The clip map's key exactly as the bin holds it: a resolved name, or
     *  `0x…` when the hashtable could not resolve it.
     *
     *  `name` substitutes the `.anm` filename stem for an unresolved key, which
     *  makes it a LABEL, not an identity — editing the animation path changed
     *  it. Rename and identity must read this instead. */
    mapKey: string;
    /** `.anm` asset path; for a composite clip, its first member's. */
    anmPath: string | null;
    /** A composite clip's ordered queue; empty for a plain atomic clip. */
    members: ClipMember[];
    /** Legacy submesh-only view; a strict subset of `allEvents`. */
    events: SubmeshVisEvent[];
    loops: boolean;
    kind: ClipKind;
    classHash: number;
    /** Raw `mFlags`, unmasked. `loops` is bit 2 of this. */
    flags: number;
    trackDataName: string | null;
    maskDataName: string | null;
    startFrame: number | null;
    endFrame: number | null;
    /** Pair list for a selector / parametric / condition-float clip. */
    pairs: ClipPair[];
    allEvents: AnimEvent[];
    addr: BinAddr;
}

/** One blend mask. `weights[i]` is the weight of skeleton joint `i` — the index
 *  IS the joint link, so never sort, filter or reorder this array. */
export interface MaskData {
    name: string;
    weights: number[];
}

/** One track. Both fields are optional in the file format; `null` means the
 *  field was absent, which is NOT the same as 0. */
export interface TrackData {
    name: string;
    priority: number | null;
    blendMode: number | null;
}

/** The `.skl` a set of masks applies to, found via the owning skin entry. */
export interface SkeletonLink {
    /** The raw `ASSETS/...` string exactly as authored in the bin. */
    sklRef: string;
    /** Resolved on-disk path, null when the project doesn't ship the file. */
    sklPath: string | null;
    sknRef: string | null;
}

export interface AnmModel {
    clips: ClipInfo[];
    masks: MaskData[];
    tracks: TrackData[];
    skeleton: SkeletonLink | null;
    /** Non-fatal problems worth showing: dangling mask / track references and
     *  mask weight counts that disagree with the skeleton's joint count. */
    warnings: string[];
}

export interface ChildParams {
    effectKey: string;
    rate: number;
    lifetime: number;
    bindWeight: number;
    translation: [number, number, number];
    isSingleParticle: boolean;
    emitterName?: string | null;
    timeBeforeFirstEmission?: number;
}

export interface PersistentPresetPayload {
    type: string;
    animationName?: string | null;
    scriptName?: string | null;
    spellHash?: string | null;
    slot?: number | null;
    operator?: number | null;
    value?: number | null;
    delayOn: number;
    delayOff: number;
}

export interface PersistentVfxPayload {
    key: string;
    boneName?: string | null;
    scale?: number | null;
    ownerOnly?: boolean | null;
    attachToCamera?: boolean | null;
    forceRender?: boolean | null;
}

export interface PersistentPayload {
    preset: PersistentPresetPayload;
    vfx: PersistentVfxPayload[];
    submeshesShow: string[];
    submeshesHide: string[];
}

export interface VfxOpenResult {
    sessionId: number;
    model: VfxPortModel;
}

export interface PortEmittersResult {
    model: VfxPortModel;
    /** Names of the emitters that were ported. */
    ported: string[];
    /** Donor asset strings to copy via portCopyAssetsToTarget. */
    assetPaths: string[];
}

export interface PortSystemResult {
    model: VfxPortModel;
    systemKey: string;
    finalName: string;
    assetPaths: string[];
}

/** Open a skin bin (and its resolvable linked bins) into a resident session. */
export function vfxOpen(path: string): Promise<VfxOpenResult> {
    return invokeCommand<VfxOpenResult>('vfx_open', { path });
}

/** Reproject the model from the live trees. */
export function vfxModel(sessionId: number): Promise<VfxPortModel> {
    return invokeCommand<VfxPortModel>('vfx_model', { sessionId });
}

/** Project the animation read layer: clips, blend masks, tracks, the skeleton
 *  those masks index into, and the non-fatal warnings worth surfacing. */
export function vfxAnmModel(sessionId: number): Promise<AnmModel> {
    return invokeCommand<AnmModel>('vfx_anm_model', { sessionId });
}

/** Reparse the session when any loaded BIN changed outside Quartz. */
export function vfxReloadIfChanged(sessionId: number): Promise<VfxPortModel | null> {
    return invokeCommand<VfxPortModel | null>('vfx_reload_if_changed', { sessionId });
}

/** Write every dirty bin back to its own file. Returns the paths written.
 *  Rejects with a `STALE_FILE:` error if a file changed on disk since opening,
 *  unless `force` is true (see {@link isStaleFileError}). */
export function vfxSave(sessionId: number, force = false): Promise<string[]> {
    return invokeCommand<string[]>('vfx_save', { sessionId, force });
}

/** Close a session and free its trees. */
export function vfxClose(sessionId: number): Promise<boolean> {
    return invokeCommand<boolean>('vfx_close', { sessionId });
}

/** Undo the last edit. Returns the refreshed model, or null if nothing to undo. */
export function vfxUndo(sessionId: number): Promise<VfxPortModel | null> {
    return invokeCommand<VfxPortModel | null>('vfx_undo', { sessionId });
}

/** Redo the last undone edit; returns the refreshed model or null. */
export function vfxRedo(sessionId: number): Promise<VfxPortModel | null> {
    return invokeCommand<VfxPortModel | null>('vfx_redo', { sessionId });
}

/** Create a new empty VFX system and register it in the resolver. */
export function vfxCreateSystem(sessionId: number, name: string): Promise<VfxPortModel> {
    return invokeCommand<VfxPortModel>('vfx_create_system', { sessionId, name });
}

/** Clone emitters from a donor session into a target system. */
export function vfxPortEmitters(
    targetSessionId: number,
    donorSessionId: number,
    donorEmitters: VfxPath[],
    targetSystem: VfxPath,
    donorGeneration?: number,
): Promise<PortEmittersResult> {
    return invokeCommand<PortEmittersResult>('vfx_port_emitters', {
        targetSessionId, donorSessionId, donorEmitters, targetSystem,
        donorGeneration: donorGeneration ?? null,
    });
}

/** Clone a whole system from a donor session into the target. */
export function vfxPortSystem(
    targetSessionId: number,
    donorSessionId: number,
    donorSystem: VfxPath,
    desiredName: string | null,
    preserveName: boolean,
    donorGeneration?: number,
): Promise<PortSystemResult> {
    return invokeCommand<PortSystemResult>('vfx_port_system', {
        targetSessionId, donorSessionId, donorSystem, desiredName, preserveName,
        donorGeneration: donorGeneration ?? null,
    });
}

/** Remove one emitter from its owning system. */
export function vfxDeleteEmitter(sessionId: number, emitter: VfxPath): Promise<VfxPortModel> {
    return invokeCommand<VfxPortModel>('vfx_delete_emitter', { sessionId, emitter });
}

/** Delete several emitters as ONE edit: a single undo step and a single model
 *  rebuild. Looping `vfxDeleteEmitter` instead costs one IPC round-trip plus a
 *  full reprojection per emitter, and makes undo restore them one by one. */
export function vfxDeleteEmitters(sessionId: number, emitters: VfxPath[]): Promise<VfxPortModel> {
    return invokeCommand<VfxPortModel>('vfx_delete_emitters', { sessionId, emitters });
}

/** Remove a system entry and any resolver entries pointing at it. */
export function vfxDeleteSystem(sessionId: number, system: VfxPath): Promise<VfxPortModel> {
    return invokeCommand<VfxPortModel>('vfx_delete_system', { sessionId, system });
}

/** Upsert (16 floats) or remove (null) a system's transform matrix. */
export function vfxSetMatrix(
    sessionId: number,
    system: VfxPath,
    values: number[] | null,
): Promise<VfxPortModel> {
    return invokeCommand<VfxPortModel>('vfx_set_matrix', { sessionId, system, values });
}

/** Add one idle-particle effect per bone for the effect key. */
export function vfxIdleAdd(
    sessionId: number,
    effectKey: string,
    bones: string[],
): Promise<VfxPortModel> {
    return invokeCommand<VfxPortModel>('vfx_idle_add', { sessionId, effectKey, bones });
}

/** Remove every idle-particle effect matching the effect key. */
export function vfxIdleRemove(sessionId: number, effectKey: string): Promise<VfxPortModel> {
    return invokeCommand<VfxPortModel>('vfx_idle_remove', { sessionId, effectKey });
}

/** Append a new child (_cbdl) emitter to the host system. */
export function vfxChildAdd(
    sessionId: number,
    hostSystem: VfxPath,
    params: ChildParams,
): Promise<VfxPortModel> {
    return invokeCommand<VfxPortModel>('vfx_child_add', { sessionId, hostSystem, params });
}

/** Overwrite an existing child emitter's parameters. */
export function vfxChildUpdate(
    sessionId: number,
    emitter: VfxPath,
    params: ChildParams,
): Promise<VfxPortModel> {
    return invokeCommand<VfxPortModel>('vfx_child_update', { sessionId, emitter, params });
}

/** Append (index = null) or replace (index = i) a persistent-effect condition. */
export function vfxPersistentUpsert(
    sessionId: number,
    index: number | null,
    payload: PersistentPayload,
): Promise<VfxPortModel> {
    return invokeCommand<VfxPortModel>('vfx_persistent_upsert', { sessionId, index, payload });
}

/** Remove one persistent-effect condition by index. */
export function vfxPersistentRemove(sessionId: number, index: number): Promise<VfxPortModel> {
    return invokeCommand<VfxPortModel>('vfx_persistent_remove', { sessionId, index });
}

/** Upsert a resolver map entry (key hash -> system link). */
export function vfxResolverUpsert(
    sessionId: number,
    key: string,
    value: string,
): Promise<VfxPortModel> {
    return invokeCommand<VfxPortModel>('vfx_resolver_upsert', { sessionId, key, value });
}

/** Set an emitter's emitterName. */
export function vfxRenameEmitter(
    sessionId: number,
    emitter: VfxPath,
    newName: string,
): Promise<VfxPortModel> {
    return invokeCommand<VfxPortModel>('vfx_rename_emitter', { sessionId, emitter, newName });
}

/** Rewrite an emitter's texture path. `oldPath` identifies the node (its
 *  current value); `newPath` replaces it. */
export function vfxSetTexture(
    sessionId: number,
    emitter: VfxPath,
    oldPath: string,
    newPath: string,
): Promise<VfxPortModel> {
    return invokeCommand<VfxPortModel>('vfx_set_texture', { sessionId, emitter, oldPath, newPath });
}

/** Rename a system: particleName/particlePath, path hash, and resolver relink. */
export function vfxRenameSystem(
    sessionId: number,
    system: VfxPath,
    newName: string,
): Promise<VfxPortModel> {
    return invokeCommand<VfxPortModel>('vfx_rename_system', { sessionId, system, newName });
}
